smarts/core/coordinates.py

```diff
@@ -118,6 +120,144 @@ def __del__(self):
             del _shapely_points[self]
 
 
+def euclidean_distance(p1: Point, p2: Point) -> float:
+    """The distance taking measuring a direct line between p1 and p2."""
+    dx = p1[0] - p2[0]
+    dy = p1[1] - p2[1]
+    return math.sqrt(dx * dx + dy * dy)
+
+
+def position_at_offset(p1: Point, p2: Point, offset: float) -> Optional[Point]:
+    """A point between p1 and p2 given an offset less than the distance between p1 and p2."""
+    if is_close(offset, 0.0):  # for pathological cases with dist == 0 and offset == 0
+        return p1
+
+    dist = euclidean_distance(p1, p2)
+
+    if is_close(dist, offset):
+        return p2
+
+    if offset > dist:
+        return None
+
+    x = p1[0] + (p2[0] - p1[0]) * (offset / dist)
+    y = p1[1] + (p2[1] - p1[1]) * (offset / dist)
+    return Point(x, y)
+
+
+def offset_along_shape(point: Point, shape: List[Point]) -> Union[float, int]:
```

Comment:
I have some concern that these methods are shape-related rather than coordinate-specific. I would like to see them in a ``core/shape.py`` rather than this file.